File: B. Medium/0947-most-stones-removed-with-same-row-or-column/0947-most-stones-removed-with-same-row-or-column.py
from typing import List

class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        num_stones = len(stones)
        # rank[i] will hold the rank (size) of the set containing stone i
        rank = [1] * num_stones
        # parent[i] will hold the parent of stone i in the union-find structure
        parent = [i for i in range(num_stones)]
        
        # Find the root parent of the stone with path compression
        def find(stone_index):
            while stone_index != parent[stone_index]:
                # Path compression: point the stone directly to its grandparent
                parent[stone_index] = parent[parent[stone_index]]
                stone_index = parent[stone_index]
            return stone_index
        
        # Union two stones' sets if they are not already connected
        def union(stone1_index, stone2_index):
            root1 = find(stone1_index)
            root2 = find(stone2_index)
            
            # If both stones are already in the same set, no need to union
            if root1 == root2:
                return 0
            
            # Union by rank: attach the smaller tree under the larger tree
            if rank[root1] < rank[root2]:
                root1, root2 = root2, root1  # Swap to ensure root1 is always larger
            parent[root2] = root1
            rank[root1] += rank[root2]
            return 1
        
        # Maps to track the latest stone seen in each row and column
        row_to_stone = {}
        col_to_stone = {}
        stones_removed = 0
        
        # Process each stone
        for i, (row, col) in enumerate(stones):
            # If there is already a stone in the same row, union the current stone with it
            if row in row_to_stone:
                stones_removed += union(i, row_to_stone[row])
            else:
                row_to_stone[row] = i
            
            # If there is already a stone in the same column, union the current stone with it
            if col in col_to_stone:
                stones_removed += union(i, col_to_stone[col])
            else:
                col_to_stone[col] = i
        
        # The number of stones that can be removed is equal to the number of unions performed
        return stones_removed

File: B. Medium/0947-most-stones-removed-with-same-row-or-column/0947-most-stones-removed-with-same-row-or-column.c
#include <stdlib.h>
#include <stdbool.h>

// Union-Find data structure
typedef struct {
    int* parent;
    int* rank;
} UnionFind;

// Function to create and initialize Union-Find structure
UnionFind* createUnionFind(int size) {
    UnionFind* uf = (UnionFind*)malloc(sizeof(UnionFind));
    uf->parent = (int*)malloc(size * sizeof(int));
    uf->rank = (int*)calloc(size, sizeof(int)); // rank is initialized to 0
    for (int i = 0; i < size; i++) {
        uf->parent[i] = i; // Initially, each element is its own parent
    }
    return uf;
}

// Function to free the Union-Find structure
void destroyUnionFind(UnionFind* uf) {
    free(uf->parent);
    free(uf->rank);
    free(uf);
}

// Find function with path compression
int find(UnionFind* uf, int x) {
    if (uf->parent[x] != x) {
        uf->parent[x] = find(uf, uf->parent[x]); // Path compression
    }
    return uf->parent[x];
}

// Union function with union by rank
void unionSets(UnionFind* uf, int x, int y) {
    int rootX = find(uf, x);
    int rootY = find(uf, y);
    if (rootX == rootY) return; // Already in the same set

    // Union by rank
    if (uf->rank[rootX] > uf->rank[rootY]) {
        uf->parent[rootY] = rootX;
    } else if (uf->rank[rootX] < uf->rank[rootY]) {
        uf->parent[rootX] = rootY;
    } else {
        uf->parent[rootY] = rootX;
        uf->rank[rootX]++;
    }
}

// Function to calculate the maximum number of stones that can be removed
int removeStones(int** stones, int stonesSize, int* stonesColSize) {
    int maxRow = 0, maxCol = 0;

    // Find maximum row and column values
    for (int i = 0; i < stonesSize; i++) {
        if (stones[i][0] > maxRow) maxRow = stones[i][0];
        if (stones[i][1] > maxCol) maxCol = stones[i][1];
    }

    int totalSize = maxRow + maxCol + 2; // +2 because we're indexing from 0
    UnionFind* uf = createUnionFind(totalSize);

    // Union stones in the same row or column
    for (int i = 0; i < stonesSize; i++) {
        unionSets(uf, stones[i][0], stones[i][1] + maxRow + 1);
    }

    int uniqueComponents = 0;
    bool* visited = (bool*)calloc(totalSize, sizeof(bool));

    // Count the number of unique connected components
    for (int i = 0; i < stonesSize; i++) {
        int root = find(uf, stones[i][0]);
        if (!visited[root]) {
            visited[root] = true;
            uniqueComponents++;
        }
    }

    free(visited);
    destroyUnionFind(uf);

    // The number of stones that can be removed
    return stonesSize - uniqueComponents;
}
